tensorflow_addons/optimizers/conditional_gradient.py

```diff
@@ -131,23 +193,31 @@ def frobenius_norm(m):
         return tf.group(var_update_op)
 
     def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
-        def frobenius_norm(m):
-            return tf.reduce_sum(m ** 2) ** 0.5
-
         var_device, var_dtype = var.device, var.dtype.base_dtype
         coefficients = (apply_state or {}).get(
             (var_device, var_dtype)
         ) or self._fallback_apply_state(var_device, var_dtype)
-        norm = tf.convert_to_tensor(
-            frobenius_norm(grad), name="norm", dtype=var.dtype.base_dtype
-        )
         lr = coefficients["learning_rate"]
         lambda_ = coefficients["lambda_"]
         epsilon = coefficients["epsilon"]
         var_slice = tf.gather(var, indices)
-        var_update_value = tf.math.multiply(var_slice, lr) - (
-            1 - lr
-        ) * lambda_ * grad / (norm + epsilon)
+        if self.ord == "fro":
+            norm = tf.convert_to_tensor(
+                self._frobenius_norm(grad), name="norm", dtype=var.dtype.base_dtype
+            )
+            var_update_value = tf.math.multiply(var_slice, lr) - (
+                1 - lr
+            ) * lambda_ * grad / (norm + epsilon)
+        else:
+            top_singular_vector = tf.convert_to_tensor(
+                self._top_singular_vector(grad),
+                name="top_singular_vector",
+                dtype=var.dtype.base_dtype,
+            )
+            var_update_value = (
+                tf.math.multiply(var_slice, lr)
+                - (1 - lr) * lambda_ * top_singular_vector
+            )
```

Comment:
The `var_update_value` in both branches have common terms:
```
tf.math.multiply(var_slice, lr) - lambda_ * (1 - lr)
```

You can rewrite this as:

```
factor = tf.math.multiply(var_slice, lr) - lambda_ * (1 - lr)

if self.ord == "fro":
  norm = tf.convert_to_tensor(...)
  var_update_value = factor * grad / (norm + eps)
else:
  top_singlar_vector = tf.convert_to_tensor(...)
  var_update_value = factor * top_singular_vector
```

Makes the code easier to read.

Reply:
This happens in `_resource_apply_dense` as well.